src/text/non_blocking/raw_text_reader.rs

```diff
@@ -721,29 +795,37 @@ impl<A: AsRef<[u8]>> IonReader for RawTextReader<A> {
     }
 
     fn step_out(&mut self) -> IonResult<()> {
-        if self.parents.is_empty() {
-            return illegal_operation(
-                "Cannot call `step_out()` when the reader is at the top level.",
-            );
-        }
+        // It is possible to fail partway through stepping out, after we've popped the parent off
+        // and started looking for an outer container's delimiter. So we first need to check our
+        // carried state and see if we're re-attempting step_out.
+        if !matches!(self.carried_state, Some(ReaderState::ContainerDelimeter)) {
+            if self.parents.is_empty() {
+                return illegal_operation(
+                    "Cannot call `step_out()` when the reader is at the top level.",
+                );
+            }
 
-        // The container we're stepping out of.
-        let parent = self.parents.last().unwrap();
+            // The container we're stepping out of.
+            let parent = self.parents.last().unwrap();
 
-        // If we're not at the end of the current container, advance the cursor until we are.
-        // Unlike the binary reader, which can skip-scan, the text reader must visit every value
-        // between its current position and the end of the container.
-        if !parent.is_exhausted() {
-            while let RawStreamItem::Value(_) | RawStreamItem::Null(_) = self.next()? {}
-        }
+            // If we're not at the end of the current container, advance the cursor until we are.
+            // Unlike the binary reader, which can skip-scan, the text reader must visit every value
+            // between its current position and the end of the container.
+            if !parent.is_exhausted() {
+                while let RawStreamItem::Value(_) | RawStreamItem::Null(_) = self.next()? {}
+            }
 
-        // Remove the parent container from the stack and clear the current value.
-        let _ = self.parents.pop();
-        self.current_value = None;
+            // Remove the parent container from the stack and clear the current value.
+            let _ = self.parents.pop();
+            self.current_value = None;
 
-        if self.parents.is_empty() {
-            // We're at the top level; nothing left to do.
-            return Ok(());
+            if self.parents.is_empty() {
+                // We're at the top level; nothing left to do.
+                return Ok(());
+            }
+            // We have stepped out and need to track that we have in just in case we error due to
+            // incomplete text while looking.
+            self.carried_state = Some(ReaderState::ContainerDelimeter);
```

Comment:
This strikes me as late in the method to set `carried_state`; what if the loop on line 815 encounters incomplete data while it's exhausting all of the child elements of the container we're stepping out of?